### DISCUSSION TITLE
Writing test classes

### DISCUSSION THREAD

**Discussion by baobao917** [jamessimone/apex-rollup#515] [Q&A]
Hi @jamessimone . This is a follow-up to #512 .

Since it is necessary to write an apex trigger for any parent objects where merging is supported (Account, Case, Contact, Lead), it would be helpful if there were more clear instructions on how to create test classes for them (geared towards an audience of non-developers, i.e. consultants, admins, adminelepors).  

One nice feature of DLRS is that when you deployed the trigger it also created the test class for you.  While I understand (from a prior thread) that this is one of the core issues with DLRS performance/limitations, one of the reasons why DLRS has such a high adoption rate is that you don't need a coding/developer background to use.

Since it is common to create rollups where the parent is one of those 4 standard objects, per #512 , it is almost always going to require an apex trigger and therefore an accompanying test class.  I've created the trigger pretty easily based on the README documentation, however creating the test class has proven to be significantly more difficult (I'm not a developer).

I have reviewed the documentation and have found the following interspersed which may or may not be relevant:
- **DEPLOYMENT & SETUP section**
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/aac26f20-5518-4ff2-896c-204b94d3afb9)

- **YOUTUBE VIDEO TUTORIAL section**
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/59fbf522-28db-4da5-ba4a-ae4e1b223109)

- **PARENT LEVEL MERGES section**
-- Although in this section the code snippet is only for an (after delete) trigger, per #512 discussion, the best practice is to have a trigger with all of the trigger contexts

![image](https://github.com/jamessimone/apex-rollup/assets/104730171/36a6eb77-35cb-4880-a0ba-77ab392de614)

-- This snippet which I'm not sure where it fits into this use case/discussion
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/cecc1604-416f-4d61-98b0-336f50a6736a)

- **CUSTOM APEX ROLLUPS section**
- This entire section appears to be geared towards developers and not sure if any of this is applicable to this use case/discussion.  I can't even begin to understand it.

**QUESTIONS:**
1. So using the information above, I created the Apex trigger (which works) and the test class (which doesn't).   What am I doing wrong?
- Account apex trigger
trigger ApexRollupAccountTrigger on Account(after insert, after update, before delete, after delete, after undelete) {
  // after delete is required ONLY if your org does Account / Contact / Lead / Case merges
  Rollup.runFromTrigger();
  // etc. You can invoke the above from your handler if you have one
  }
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/10d13e74-19e9-4337-b5d3-42417e3c1f02)

- Test class (with screenshot of error)
@IsTest
private class ApexRollupAccountTest
{
@TestSetup
static void setup() {
  upsert new RollupSettings__c(IsEnabled__c = true);

// —— DATA PREPARATION ——
         // insert 1 Account
         Account acct = new Account(
                 Name = 'Test Account',
                 Tax_Code__c = 'ICOU78');
         insert acct;
 
// Force the ApexRollupAccountTrigger to be invoked, fails the test if org config or other Apex code prevents this.
Test.startTest();
        ApexRollupAccountTrigger(new Account());
Test.stopTest();
    }
 }
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/b6d4c56d-7304-4fc8-9001-36b9f09d1d6b)

**SUGGESTIONS:**
1. Create a new section specifically for test classes.  Consolidate all the information interspersed in this section. And step by step instructions.
2. YouTube video was basically too hard to follow along and was too geared towards a developer audience.  Could not comprehend much of the topics.
3. What I found helpful for the apex trigger, was the basic code template which a user could then modify. (See screenshot below).  Since it doesn't seem possible to automate creation of triggers and test classes (like in DLRS) and has to be done manually, it would be extremely helpful to include similar templates for test classes on these objects.
![image](https://github.com/jamessimone/apex-rollup/assets/104730171/bb0ba8eb-17c4-46c3-ad39-e8e98527e7c2)

**Comment by jamessimone** [comment]
your test class should look like this:

```java
@IsTest
private class ApexRollupAccountTest {
  @TestSetup
  static void setup() {
    upsert new RollupSettings__c(IsEnabled__c = true);
  }

  @IsTest
  static void triggerIntegratesSuccessfully() {
    Test.startTest();
    insert new Account(
      Name = 'Test Account',
      Tax_Code__c = 'ICOU78'
    );
    Test.stopTest();
    
    Assert.areEqual(0, Rollup.CACHED_ROLLUPS.size());
  }
}
```

There's a lot of information to digest; I totally get it. I will have to go through your feedback here and see what I'd like to incorporate.

**Comment by baobao917** [comment]
Hi @jamessimone - I saw in v.1.6.0 release notes that the README was updated pertaining to this.  Can you just let me know (or point me to where) what changes were incorporated so I can review?

**Comment by jamessimone** [comment]
@baobao917 sorry for the confusion. Here's what I was referencing:

> This snippet which I'm not sure where it fits into this use case/discussion

I removed that snippet, as it was out of date. I am still planning to further update the README and wiki with a more thorough testing baseline example based on the rest of your feedback above.
